### PR TITLE
backend/remote: Swift Authentication Update

### PR DESCRIPTION
**Pull request by kayrus**
This PR syncs OpenStack auth options with provider auth options.

See https://github.com/terraform-providers/terraform-provider-openstack/issues/686 for a reference.

/cc @jtopjian @radeksimko

### CODE REVIEW

**Comment by SvenDowideit** [comment]
would this PR mean that 

```
provider "openstack" {
 user_name = data.vault_generic_secret.nectarcloud_ereefs.data["openstack_user_name"]
 tenant_id = data.vault_generic_secret.nectarcloud_ereefs.data["openstack_tenant_id"]
 tenant_name = data.vault_generic_secret.nectarcloud_ereefs.data["openstack_tenant_name"]
 password = data.vault_generic_secret.nectarcloud_ereefs.data["openstack_password"]
 auth_url = data.vault_generic_secret.nectarcloud_ereefs.data["openstack_auth_url"]
}

terraform {
  backend "swift" {
    container         = "terraform-state"
    archive_container = "terraform-state-archive"
  }
}
```

would just work?

ie https://github.com/terraform-providers/terraform-provider-openstack/issues/939 and https://github.com/terraform-providers/terraform-provider-openstack/issues/373

**Comment by LorbusChris** [comment]
@kayrus do you mind rebasing?

**Comment by kayrus** [comment]
@LorbusChris rebased

**Comment by LorbusChris** [comment]
@kayrus thanks! maybe sqash the last two commits, though?

**Comment by kayrus** [comment]
@LorbusChris they can be squashed via github merge menu.

**Comment by kayrus** [comment]
@SvenDowideit unfortunately no.

**Comment by LorbusChris** [comment]
@radeksimko friendly ping. If you could have a look at this soon (and subsequently at https://github.com/hashicorp/terraform/pull/23601), that would be great! :)

**Comment by LorbusChris** [comment]
any update on this?

**Comment by kayrus** [comment]
@apparentlymart @radeksimko kindly ping

**Comment by LorbusChris** [comment]
ping

**Comment by LorbusChris** [comment]
this would need another rebase :/ any chance this could go in soon?

**Comment by kayrus** [comment]
rebased

**Comment by LorbusChris** [comment]
@apparentlymart ping

**Comment by LorbusChris** [comment]
@apparentlymart @radeksimko kind ping

**Comment by kayrus** [comment]
@apparentlymart @radeksimko kind ping
/cc @jtopjian @ozerovandrei could you please help somehow to pay attention to this PR? From a contributor perspective current situation looks bizarre. I rebased this PR already 6 times.

**Comment by danieldreier** [comment]
@kayrus I apologize for the need to do multiple rebases here. I know that getting updates to backends merged has been tough, and @pkolyvas (the Terraform Core product manager) and I (the Terraform Core engineering manager) are trying to improve that. The root issue has been that nobody on the core team has deep domain knowledge of Swift, so doing really good reviews of PRs is fairly challenging. We've had to individually reach out and ask favors of engineers who have that knowledge, which slows down the process a lot.

I'm hopeful that we can significantly speed up the backend review process by allowing people who contribute to this codebase to review and approve PRs. We recently updated our [contributor guide](https://github.com/hashicorp/terraform/blob/master/.github/CONTRIBUTING.md#maintainers) and set up a [CODEOWNERS](https://github.com/hashicorp/terraform/blob/master/CODEOWNERS) section to track who is maintaining what. Swift currently [lacks a maintainer](https://github.com/hashicorp/terraform/blob/master/CODEOWNERS#L17). Once we have a few maintainers listed for this backend, I'm hopeful that the community of Terraform OpenStack contributors can review, test, and approve changes, so that Core feels more confident merging these changes quickly.

If anyone on this thread would like to become a maintainer, please either make a PR to [CODEOWNERS](https://github.com/hashicorp/terraform/blob/master/CODEOWNERS) proposing adding yourself, or email me (ddreier@hashicorp.com) and Petros (petros@hashicorp.com) with any questions you have. I'm happy to email or video conference.

**Comment by kayrus** [comment]
Dear @danieldreier, thank you for your reply and explanation. This PR doesn't change the Swift logic and reviewers don't need to have a Swift knowledge. I'd prefer to avoid decisions regarding a maintainer role if possible, since this PR just synchronizes auth options between upstream terraform-provider-openstack and resolves a https://github.com/terraform-providers/terraform-provider-openstack/issues/686 opened by @radeksimko. Please let me know what do you think.

**Comment by jtopjian** [comment]
@danieldreier

As a previous maintainer of the OpenStack plugin and as someone who used to work on and review this part of the code in this repository, I can confidently say that the Swift-specific changes being done in this PR are safe.

This PR is a result of a request that was made at https://github.com/terraform-providers/terraform-provider-openstack/issues/686. The requested wanted to decouple the authentication logic being shared between the openstack plugin and this Swift remote backend:

> hashicorp/terraform also depends on terraform-providers/terraform-provider-openstack because it shares some auth logic in the Swift remote backend.

Sharing the auth logic was problematic because this caused `hashicorp/terraform` to require all of the same Go dependencies as `terraform-provider-openstack`. 

Work was done to satisfy this decoupling by creating a small package located at `github.com/gophercloud/utils/terraform/auth`. This package shared no dependencies with `terraform-provider-openstack` and thus would help remove the unnecessary dependencies from `hashicorp/terraform`. This PR is large, yes, but that is due to the amount of dependencies no longer requested as a result of decoupling

There are some changes being made to the actual Swift logic, but they are all amendments to enhance the current authentication to reach feature parity with the authentication work that has happened since https://github.com/terraform-providers/terraform-provider-openstack/issues/686 was made - if that makes sense?

Basically, https://github.com/terraform-providers/terraform-provider-openstack/issues/686 was opened, we did a bunch of work to resolve that request, this PR was opened, and then in the meantime, additional authentication amendments were made which has to also be done here in order to use the shared authentication package we created.

Again, from reviewing this PR, I feel the changes being made safe. However, if any issues are caused by this PR, I will work with @kayrus on resolving them. Unfortunately I don't have the time to become a full maintainer (I used to), but to help move this PR along, I can commit to ensuring no issues come out of this PR.

**Comment by LorbusChris** [comment]
:shipit:

**Comment by danieldreier** [comment]
@kayrus @jtopjian thank you for the extended review, thoughtful responses, and the work on contributing this and getting it merged. We'll merge this and include it in the upcoming 0.13 release. I think it makes sense to have discussions about maintainers elsewhere. We'll pin a thread in the [terraform community forum](https://discuss.hashicorp.com/c/terraform-core) and continue that discussion over there.

**Comment by pedrokiefer** [comment]
Any chance this get cherry-picked for 0.12.XX release?

**Comment by danieldreier** [comment]
[Terraform 0.13.0 beta 1 launched today](https://github.com/hashicorp/terraform/issues/25016) with this update! If you have time it would be awesome if you could try it out and make sure it works.
